edb/edgeql/compiler/setgen.py

```diff
@@ -411,10 +411,28 @@ def compile_path(expr: qlast.Path, *, ctx: context.ContextLevel) -> irast.Set:
                 raise RuntimeError(
                     'unexpected expression as a non-first path item')
 
-            with ctx.newscope(fenced=True, temporary=False) as subctx:
+            # We need to fence this if the head is a mutating
+            # statement, to make sure that the factoring allowlist
+            # works right.
+            is_subquery = isinstance(step, qlast.MutatingStatement)
+            with ctx.newscope(fenced=is_subquery) as subctx:
                 path_tip = ensure_set(
                     dispatch.compile(step, ctx=subctx), ctx=subctx)
 
+                # If the head of the path is a direct object
+                # reference, wrap it in an expression set to give it a
+                # new path id. This prevents the object path from being
+                # spuriously visible to computable paths defined in a shape
+                # at the root of a path. (See test_edgeql_select_tvariant_04
+                # for an example).
+                if (
+                    path_tip.path_id.is_objtype_path()
```

Comment:
You probably also want to add `and not path_id.is_view_path()` to actually test that `path_id` is a direct reference, as `is_objtype_path` only tests if `path_id` represents an object type.